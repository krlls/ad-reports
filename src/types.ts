export type ServerConfig = {
  port: number,
}

export namespace Posts {
  export type Post = {
    id: number,
    title: string,
    data: string,
  }

  export namespace List {
    export type Req = {
      id: number,
    }
  }
}
