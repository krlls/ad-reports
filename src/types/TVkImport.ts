export namespace IVkApi {
  export namespace Wall {
    export interface Response {
      count: number,
      items: Item[],
    }

    export interface Item {
      id: number,
      from_id: number,
      owner_id: number,
      date: number,
      marked_as_ads: number,
      can_delete: number,
      is_favorite: boolean,
      post_type: PostType,
      text: string,
      can_edit?: number,
      created_by?: number,
      can_pin: number,
      is_pinned?: number,
      attachments: Attachment[],
      post_source: PostSource,
      comments: Comments,
      likes: Likes,
      reposts: Reposts,
      views: Views,
      donut: Donut,
      short_text_rate: number,
      carousel_offset?: number,
      hash: string,
      edited?: number,
      postponed_id?: number,
    }

    export interface Attachment {
      type: AttachmentType,
      market_album?: MarketAlbum,
      link?: Link,
      video?: Video,
      photo?: LinkPhoto,
      poll?: Poll,
    }

    export interface Link {
      url: string,
      title: Title,
      caption?: string,
      description: string,
      photo?: LinkPhoto,
      is_favorite: boolean,
      target?: string,
    }

    export interface LinkPhoto {
      album_id: number,
      date: number,
      id: number,
      owner_id: number,
      sizes: Size[],
      text: string,
      user_id: number,
      has_tags: boolean,
      access_key?: string,
      post_id?: number,
    }

    export interface Size {
      height: number,
      url: string,
      type?: SizeType,
      width: number,
      with_padding?: number,
    }

    export enum SizeType {
      M = 'm',
      O = 'o',
      P = 'p',
      Q = 'q',
      R = 'r',
      S = 's',
      W = 'w',
      X = 'x',
      Y = 'y',
      Z = 'z',
    }

    export enum Title {
      Empty = '',
      LevshaKaluga = 'Levsha Kaluga',
    }

    export interface MarketAlbum {
      id: number,
      owner_id: number,
      title: string,
      count: number,
      updated_time: number,
      is_main: boolean,
      is_hidden: boolean,
      photo: LinkPhoto,
    }

    export interface Poll {
      multiple: boolean,
      end_date: number,
      closed: boolean,
      is_board: boolean,
      can_edit: boolean,
      can_vote: boolean,
      can_report: boolean,
      can_share: boolean,
      created: number,
      id: number,
      owner_id: number,
      question: string,
      votes: number,
      disable_unvote: boolean,
      anonymous: boolean,
      answer_ids: number[],
      embed_hash: string,
      photo?: PollPhoto,
      answers: Answer[],
      author_id: number,
    }

    export interface Answer {
      id: number,
      rate: number,
      text: string,
      votes: number,
    }

    export interface PollPhoto {
      color: string,
      id: number,
      images: Size[],
    }

    export enum AttachmentType {
      Link = 'link',
      MarketAlbum = 'market_album',
      Photo = 'photo',
      Poll = 'poll',
      Video = 'video',
    }

    export interface Video {
      access_key?: string,
      can_comment: number,
      can_edit: number,
      can_like: number,
      can_repost: number,
      can_subscribe: number,
      can_add_to_faves: number,
      can_add: number,
      can_attach_link?: number,
      comments: number,
      date: number,
      description?: string,
      duration: number,
      image: Size[],
      first_frame?: Size[],
      width: number,
      height: number,
      id: number,
      owner_id: number,
      ov_id?: string,
      title: string,
      is_favorite: boolean,
      track_code: string,
      type: VideoType,
      views: number,
      live_status?: string,
      is_author?: boolean,
      repeat?: number,
      content_restricted?: number,
      content_restricted_message?: string,
    }

    export enum VideoType {
      ShortVideo = 'short_video',
      Video = 'video',
    }

    export interface Comments {
      can_post: number,
      can_close: number,
      count: number,
      groups_can_post: boolean,
    }

    export interface Donut {
      is_donut: boolean,
    }

    export interface Likes {
      can_like: number,
      count: number,
      user_likes: number,
      can_publish: number,
    }

    export interface PostSource {
      type: PostSourceType,
      platform?: string,
    }

    export enum PostSourceType {
      API = 'api',
      Vk = 'vk',
    }

    export enum PostType {
      Post = 'post',
    }

    export interface Reposts {
      count: number,
      wall_count: number,
      mail_count: number,
      user_reposted: number,
    }

    export interface Views {
      count: number,
    }
  }
}
