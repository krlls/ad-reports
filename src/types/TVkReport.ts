export interface IVKReport {
  generatePostReport(): Promise<boolean>,
}

export interface VKPostReport {
  postId: string,
  fromId: string,
  ownerId: string,
  date: string,
  reposts: number,
  likes: number,
  comments: number,
  hide: number,
  joinGroup: number,
  links: number,
  reachSubscribers: number,
  reachTotal: number,
  reachViral: number,
  reachAds: number,
  report: number,
  toGroup: number,
  unsubscribe: number,
}
