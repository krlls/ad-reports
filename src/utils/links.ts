export const createPatch = (...args: string[]) => args.join('')

export const toVkId = (id: string | number) => `-${id}`
