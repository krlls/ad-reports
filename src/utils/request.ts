import { chunk } from 'lodash'

const REQ_DELAY = 500

const delay = (clb: VoidFunction, delay: number) =>
  new Promise((resolve) =>
    setTimeout(() => {
      clb()
      resolve(undefined)
    }, delay),
  )

export const chunkRequest = async <A, B>(req: (data: B[]) => A, data: B[], chunkSize: number) => {
  const chunks = chunk(data, chunkSize)
  const requests: A[] = []

  for (const param of chunks) {
    await delay(() => requests.push(req(param)), REQ_DELAY)
  }

  return Promise.all(requests)
}
