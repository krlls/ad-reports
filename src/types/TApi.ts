import { IVkApi } from './TVkImport'

export namespace Api {
  export namespace Vk {
    export const PREFIX = '/vk'

    export namespace AuthLink {
      export const URL = '/auth'
    }

    export namespace SaveToken {
      export const URL = '/token'

      export type Req = {
        access_token: string,
        expires_in: string,
        user_id: string,
      }
    }

    export namespace Posts {
      export const URL = '/posts'

      export type Req = {
        groupId: number,
      }

      export type Resp = IVkApi.Wall.Response
    }
  }
}
