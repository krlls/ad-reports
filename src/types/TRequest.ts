import { IVkApi } from './TVkImport'

export namespace Request {
  export interface BaseParams {
    access_token: string,
    v: string,
  }

  export namespace Posts {
    export namespace GetPosts {
      export const URL = 'wall.get'
      export const METHOD = 'GET'

      export enum EFilter {
        SUGGESTS = 'suggests',
        POSTPONED = 'postponed',
        OWNER = 'owner',
        OTHERS = 'others',
        ALL = 'all',
      }

      export interface Params extends BaseParams {
        owner_id: string,
        filter?: EFilter,
        domain?: string,
        offset?: number,
        count?: number,
      }

      export type Resp = {
        response: IVkApi.Wall.Response,
      }
    }
  }
}
