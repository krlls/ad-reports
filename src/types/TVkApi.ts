import { Request } from './TRequest'

export abstract class VkImport {
  abstract api: Request.Vk.Api
  abstract posts(groupId: string): Promise<VkPost[]>
  // abstract postsStats(groupId: string): any
}

export interface VkPost {
  id: string,
  fromId: string,
  ownerId: string,
  date: string,
}
