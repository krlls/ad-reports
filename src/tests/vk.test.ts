import { expect, test, describe } from '@jest/globals'

import { Api } from '../types/TApi'
import { createPatch } from '../utils/links'
import { TestApp } from './index.test'

describe('VK tests', () => {
  const vkPatch: (...args: string[]) => string = createPatch.bind(null, Api.Vk.PREFIX)

  test('Get link works', async () => {
    const response = await TestApp.get(vkPatch(Api.Vk.AuthLink.URL))

    expect(response.status).toBe(200)
    expect(response.text).toMatchSnapshot()
  })

  test('Save token', async () => {
    const response = await TestApp.get(vkPatch(Api.Vk.SaveToken.URL)).query({
      access_token: '533bacf01e11f55b536a565b57531ad114461ae8736d6506a3',
      expires_in: 9479238,
      user_id: 234,
    })

    expect(response.status).toBe(200)
  })
})
