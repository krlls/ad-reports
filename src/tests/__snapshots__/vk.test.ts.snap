// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VK tests Get link works 1`] = `
"
    <html lang="en">
      <body>
         <div style="width: 100%; height: 100%; display: flex; align-items: center; justify-content: center;">
            <h1><a target="_blank" href="https://oauth.vk.com/authorize?client_id=51467723
&display=page
&redirect_uri=http://localhost:3000/vk/token/
&scope=wall,stats,groups
&response_type=token
&v=5.131
">Vk Auth link</a></h1>
         </div>
      </body>
    </html>
"
`;
